src/wp-includes/theme.php

```diff
@@ -322,23 +333,31 @@ function get_template() {
  *
  * @since 1.5.0
  *
+ * @global string $wp_template_path Current theme template directory path.
+ *
  * @return string Path to active theme's template directory.
  */
 function get_template_directory() {
-	$template     = get_template();
-	$theme_root   = get_theme_root( $template );
-	$template_dir = "$theme_root/$template";
+	global $wp_template_path;
 
-	/**
-	 * Filters the active theme directory path.
-	 *
-	 * @since 1.5.0
-	 *
-	 * @param string $template_dir The path of the active theme directory.
-	 * @param string $template     Directory name of the active theme.
-	 * @param string $theme_root   Absolute path to the themes directory.
-	 */
-	return apply_filters( 'template_directory', $template_dir, $template, $theme_root );
+	if ( null === $wp_template_path ) {
+		$template     = get_template();
+		$theme_root   = get_theme_root( $template );
+		$template_dir = "$theme_root/$template";
+
+		/**
+		 * Filters the active theme directory path.
+		 *
+		 * @since 1.5.0
+		 *
+		 * @param string $template_dir The path of the active theme directory.
+		 * @param string $template     Directory name of the active theme.
+		 * @param string $theme_root   Absolute path to the themes directory.
+		 */
+		$wp_template_path = apply_filters( 'template_directory', $template_dir, $template, $theme_root );
```

Comment:
See https://github.com/WordPress/wordpress-develop/pull/5160/files#r1319822198

Reply:
See my reply in https://github.com/WordPress/wordpress-develop/pull/5160#discussion_r1320186178, I think we can keep this as is, and mark the ticket to need a dev note.